composer/trainer/trainer.py

```diff
@@ -439,6 +441,25 @@ def __init__(
     def deepspeed_enabled(self):
         return self.deepspeed_config is not None
 
+    @property
+    def saved_checkpoint_filepaths(self):
```

Comment:
I'm hesitant about adding this logic into the trainer itself, as a) the trainer should not know about what callbacks are being used, b) all the other checkpointing logic lives in the checkpointer class, and c) it would not be possible for the user to differentiate between getting a local path or the remote path. In addition, these this as written would not work for multi-rank training, since each rank would have it's own checkpoint saver.

Intead, would it work to split this functionality across the checkpoint saver and run directory uploader?

I'm thinking that the checkpoint saver could have a method that returns the paths to all checkpoints. It would perform a `dist.all_gather` to gather any checkpoint files from other ranks, and then return this as a dictionary of {timestamp: list of checkpoint files for that timestamp}. For example:
```python3
def get_saved_checkpoints(self) -> Dict[Timestamp, List[pathlib.Path]]
   """Returns a list of (timestamp, List[filepath]) tuples corresponding to when a checkpoint was saved and all files from all ranks corresponding to this checkpoint. This function performs a distributed reduction to gather all checkpoint files from all ranks"""
   ...
```

Then, could the run directory uploader have a method like this?

```python3
def get_uri_for_uploaded_file(self, local_filepath: Union[pathlib.Path, str]) -> str:
    """"Returns the URI to where ``local_filepath`` was uploaded or would be uploaded. ``local_filepath`` must be in the run directory, or a ValueError would be raised"""
    ...
```

Then, the user would be able to:
```python3

run_directory_uploader = RunDirectoryUploader(...)
trainer = Trainer(callbacks=[run_directory_uploader], checkpoint_folder="checkpoints")
trainer.fit()
for checkpoint_timestamp, checkpoint_filepaths in self.checkpoint_saver.get_saved_checkpoints():
    print("Local filepaths", checkpoint_filepaths)
    print("Uploaded filepaths", [run_directory_uploader.get_uri_for_uploaded_file(x) for x in checkpoint_filepaths])
```

Reply:
Yup this sounds good!